CharlieBackend.Data/Repositories/Impl/DashboardRepository.cs

```diff
@@ -0,0 +1,338 @@
+                      / (double)x.Count() * 100)
+                 }).ToList();
+
+            return studentsAverageVisitsList;
+        }
+
+        public async Task<List<StudentVisitDto>> GetStudentsPresenceListByStudentIds(IEnumerable<long> studentIds)
+        {
+            var studentsPresenceList = await _applicationContext.Visits
+                    .AsNoTracking()
+                    .Where(x => x.Lesson.StudentGroup.StudentsOfStudentGroups
+                    .Any(x => studentIds.Contains(x.Student.Id)))
+                    .Select(x => new StudentVisitDto
+                    {
+                        CourseId = x.Lesson.StudentGroup.CourseId,
+                        StudentGroupId = (long)x.Lesson.StudentGroupId,
+                        StudentId = (long)x.StudentId,
+                        LessonId = x.LessonId,
+                        LessonDate = x.Lesson.LessonDate,
+                        Presence = x.Presence,
+                    }).ToListAsync();
+
+            return studentsPresenceList;
+        }
+
+        public async Task<List<StudentMarkDto>> GetStudentsMarksListByStudentIds(IEnumerable<long> studentIds)
+        {
+            var studentsMarksList = await _applicationContext.Visits
+                    .AsNoTracking()
+                    .Where(x => x.Lesson.StudentGroup.StudentsOfStudentGroups
+                    .Any(x => studentIds.Contains(x.Student.Id)) 
+                    && x.StudentMark != null)
+                    .Select(x => new StudentMarkDto
+                    {
+                        CourseId = x.Lesson.StudentGroup.CourseId,
+                        StudentGroupId = (long)x.Lesson.StudentGroupId,
+                        StudentId = (long)x.StudentId,
+                        LessonId = x.LessonId,
+                        LessonDate = x.Lesson.LessonDate,
+                        StudentMark = x.StudentMark,
+                    }).ToListAsync();
+
+            return studentsMarksList;
+        }
+
+        public async Task<List<long>> GetGroupsIdsByStudentIdAndPeriodAsync(long studentId, 
+            DateTime? startDate, DateTime? finishDate)
+        {
+                var groupIdsbyStudentIdAndPeriod = await _applicationContext.StudentGroups
+                    .AsNoTracking()
+                    .Where(x => x.StudentsOfStudentGroups.Any(x => x.StudentId == studentId))
+                    .WhereIf(startDate != null && finishDate != default(DateTime), 
+                    x => x.StartDate >= startDate)
+                    .WhereIf(finishDate != null && finishDate != default(DateTime), 
+                    x => x.FinishDate <= finishDate)
+                    .Select(x => x.Id)
+                    .ToListAsync();
+
+                return groupIdsbyStudentIdAndPeriod;
+        }
+
+        public async Task<List<AverageStudentMarkDto>> GetStudentAverageMarksByStudentIdAsync(long studentId, List<long> studentGroupsIds)
```

Comment:
This method has no difference with GetStudentsAverageMarksByStudentIdsAsync. please make common private method and pass required where Expression there, to somehow decrease code duplication.